atomic_reactor/plugins/pre_pin_operator_digest.py

```diff
@@ -50,25 +53,188 @@ def __init__(self, tasker, workflow, replacement_pullspecs=None):
+        self.log.info("Looking for operator CSV files in %s", manifests_dir)
+
+        operator_manifest = OperatorManifest.from_directory(manifests_dir)
+
+        if operator_manifest.files:
+            # TODO: log paths in a particular order? make sure OperatorManifest orders them?
+            path_lines = "\n".join(f.path for f in operator_manifest.files)
+            self.log.info("Found operator CSV files:\n%s", path_lines)
+        else:
+            self.log.info("No operator CSV files found")
+
+        return operator_manifest
+
+    def _get_pullspecs(self, operator_manifest):
+        pullspecs = set()
+        for operator_csv in operator_manifest.files:
+            pullspecs.update(operator_csv.get_pullspecs())
+
+        if pullspecs:
+            pullspec_lines = "\n".join(sorted(map(str, pullspecs)))
+            self.log.info("Found pullspecs:\n%s", pullspec_lines)
+        else:
+            self.log.info("No pullspecs found")
+
+        return pullspecs
+
+    def _get_replacement_pullspecs(self, pullspecs):
+        self.log.info("Computing replacement pullspecs")
+        replacer = PullspecReplacer(user_config=self.user_config, site_config=self.site_config)
+
+        for p in pullspecs:
+            if not replacer.registry_is_allowed(p):
+                raise RuntimeError("Registry not allowed: {} (in {})".format(p.registry, p))
+
+        replacements = {}
+
+        for original in pullspecs:
+            pinned = replacer.pin_digest(original)
+            replaced = replacer.replace_registry(pinned)
+            if replaced != original:
+                replacements[original] = replaced
+
+        replacement_lines = "\n".join(
+            "{} -> {}".format(p, replacements[p]) if p in replacements
+            else "{} - no change".format(p)
+            for p in sorted(pullspecs, key=str)  # ImageName does not implement ordering
+        )
+        self.log.info("To be replaced:\n%s", replacement_lines)
+
+        return replacements
+
+    def _set_worker_arg(self, replacement_pullspecs):
+        arg = {str(old): str(new) for old, new in replacement_pullspecs.items()}
+        override_build_kwarg(self.workflow, "operator_bundle_replacement_pullspecs", arg)
+
+
+_KEEP = object()
+
+
+class PullspecReplacer(object):
+    """
+    Helper that takes care of replacing parts of image pullspecs
+    """
+
+    def __init__(self, user_config, site_config):
+        """
+        Initialize a PullspecReplacer
+
+        :param user_config: container.yaml operator_manifest configuration
+        :param site_config: reactor-config-map operator_manifests configuration
+        """
+        self.allowed_registries = site_config["allowed_registries"]
+        if self.allowed_registries is not None:
+            self.allowed_registries = set(self.allowed_registries)
+
+        # TODO: what if a replacement registry is not in allowed_registries?
```

Comment:
It's fine then (but weird). There may be use case to prevent maintainer use some target registries in manifests.

Reply:
Sounds reasonable :+1: